packages/transactions/src/OutboundShipment/DetailView/reducer.test.ts

```diff
@@ -0,0 +1,228 @@
+      state,
+      OutboundAction.onSortBy(quantityColumn)
+    );
+
+    expect(reducerResult.draft.lines).toEqual(
+      [
+        expect.objectContaining({ id: '1' }),
+        expect.objectContaining({ id: '1' }),
+        expect.objectContaining({ id: '2' }),
+        expect.objectContaining({ id: '3' }),
+        expect.objectContaining({ id: '4' }),
+        expect.objectContaining({ id: '5' }),
+      ].reverse()
+    );
+  });
+
+  it('sorts the lines by the provided key in ascending order when sorted by some other key.', () => {
+    const state: OutboundShipmentStateShape = {
+      draft: { ...placeholderTransaction, lines },
+      sortBy: { key: 'quantity', isDesc: true, direction: 'desc' },
+    };
+    const { result } = renderHook(
+      () =>
+        useColumns(
+          new ColumnSetBuilder<ItemRow>().addColumn('itemName').build()
+        ),
+      { wrapper: TestingProvider }
+    );
+
+    const itemNameColumn = result.current[0];
+
+    if (!itemNameColumn) throw new Error('This test is broken!');
+
+    const reducerResult = reducer(undefined, null)(
+      state,
+      OutboundAction.onSortBy(itemNameColumn)
+    );
+
+    expect(reducerResult.draft.lines).toEqual([
+      expect.objectContaining({ id: '1' }),
+      expect.objectContaining({ id: '5' }),
+      expect.objectContaining({ id: '3' }),
+      expect.objectContaining({ id: '1' }),
+      expect.objectContaining({ id: '2' }),
+      expect.objectContaining({ id: '4' }),
+    ]);
+  });
+
+  it('updates the correct line with the correct quantity', () => {
+    const state: OutboundShipmentStateShape = {
+      draft: { ...placeholderTransaction, lines },
+      sortBy: { key: 'quantity', isDesc: true, direction: 'desc' },
+    };
+
+    const reducerResult = reducer(undefined, null)(
+      state,
+      OutboundAction.updateQuantity('1', 10)
+    );
+
+    const line = reducerResult.draft.lines.find(({ id }) => id === '1');
+
+    if (!line) throw new Error('This test is broken!');
+
+    expect(line.quantity).toBe(10);
+  });
+
+  it('updates the client side line state by merging the server data into the client data lines, where the server data always wins', () => {
+    const state: OutboundShipmentStateShape = {
+      draft: { ...placeholderTransaction, lines },
+      sortBy: { key: 'quantity', isDesc: true, direction: 'desc' },
+    };
+
+    // Create some server data which is the same except every line has 99 quantity.
+    // Then after merging, every line should have 99 quantity.
+    const dataLines = lines.map(line => ({ ...line, quantity: 99 }));
+    const data: Transaction = { ...placeholderTransaction, lines: dataLines };
+
+    const reducerResult = reducer(data, null)(state, DocumentAction.merge());
+
+    // Check for any lines that don't have a quantity of 99. If there are any, the merge was wrong.
+    expect(
+      reducerResult.draft.lines.filter(({ quantity }) => quantity !== 99).length
```

Comment:
Probably a more comprehensive 'merging' might be needed. But this test I think is good for showing the feature. For example, we might not want the quantity to be overridden but instead have a `clientSideEnteredValueForQuantityHere` and a `quantity` to show 'you entered X but it wasn't correctly updated on the server for Y reason, so we've changed it to  Y' kind of thing.

Additionally need to handle other people deleting things etc (ideally not just removing the line a user might be editing at the time..?)

Reply:
yeah, would be neat to tell people that someone else deleted a line